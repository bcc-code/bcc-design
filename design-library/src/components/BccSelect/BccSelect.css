@layer components {
    .bcc-select {
        @apply w-full h-10 px-4 py-2 form-select text-body-small rounded-lg text-primary bg-primary border-on-primary focus:ring-0 focus:outline-2 focus:outline-emphasis focus:-outline-offset-1;
        @apply disabled:bg-secondary disabled:text-tertiary disabled:cursor-not-allowed;
    }

    /* Size */
    .bcc-select-sm {
        @apply h-8 text-body-small py-1;
    }
    .bcc-select-lg {
        @apply h-14 text-body-base py-4;
    }

    /* State */
    .bcc-select-error {
        @apply border-danger focus:outline-danger;
    }
    .bcc-select-success {
        @apply border-success focus:outline-success;
    }
}
