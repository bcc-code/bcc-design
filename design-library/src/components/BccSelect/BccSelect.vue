<script lang="ts">
export default {
  inheritAttrs: false,
};
</script>

<script setup lang="ts">
import { computed, type StyleValue } from "vue";
import { useAttrsWithoutStyles } from "@/composables/attrsWithoutStyles";
import { useId } from "@/hooks/use-id";
import BccFormLabel from "@/components/BccFormLabel/BccFormLabel.vue";
import BccFormMessage from "@/components/BccFormMessage/BccFormMessage.vue";

type Props = {
  modelValue?: string;
  state?: "default" | "error" | "success";
  size?: "sm" | "base" | "lg";
  disabled?: boolean;
  label?: string;
  showOptionalLabel?: boolean;
  optionalLabel?: string;
  required?: boolean;
};

const props = withDefaults(defineProps<Props>(), {
  state: "default",
  size: "base",
  disabled: false,
  required: false,
  showOptionalLabel: false,
  optionalLabel: "Optional",
});

const id = `bcc-input-${useId()}`;

const emit = defineEmits(["update:modelValue"]);

const showOptionalLabel = computed(() => props.showOptionalLabel && !props.required);

const { attrsWithoutStyles } = useAttrsWithoutStyles();
</script>

<template>
  <div class="bcc-input-container" :class="$attrs['class']" :style="$attrs['style'] as StyleValue">
    <BccFormLabel
      v-if="label || showOptionalLabel"
      :for="id"
      :size="size === 'lg' ? 'lg' : undefined"
      :showOptionalLabel="showOptionalLabel"
      :optionalLabel="optionalLabel"
    >
      {{ label }}
    </BccFormLabel>
    <select
      class="bcc-select"
      :class="{
        'bcc-select-error': state === 'error',
        'bcc-select-success': state === 'success',
        'bcc-select-sm': size === 'sm',
        'bcc-select-lg': size === 'lg',
      }"
      :id="id"
      :value="modelValue"
      :disabled="disabled"
      :required="required"
      @input="emit('update:modelValue', ($event.target as HTMLSelectElement).value)"
      v-bind="attrsWithoutStyles"
    >
      <slot />
    </select>

    <BccFormMessage v-if="$slots.message" :state="state">
      <slot name="message" />
    </BccFormMessage>
  </div>
</template>
