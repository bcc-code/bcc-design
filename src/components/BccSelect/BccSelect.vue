<script setup lang="ts">
import { computed, type StyleValue } from "vue";
import { useAttrsWithoutStyles } from "@/composables/attrsWithoutStyles";
import { useId } from "@/hooks/use-id";

type Props = {
  modelValue?: string;
  state?: "default" | "error" | "success";
  size?: "sm" | "base" | "lg";
  disabled?: boolean;
  label?: string;
  showOptionalLabel?: boolean;
  optionalLabel?: string;
  required?: boolean;
};

const props = withDefaults(defineProps<Props>(), {
  state: "default",
  size: "base",
  disabled: false,
  required: false,
  showOptionalLabel: false,
  optionalLabel: "Optional",
});

const id = `bcc-input-${useId()}`;

const emit = defineEmits(["update:modelValue"]);

const showOptionalLabel = computed(() => props.showOptionalLabel && !props.required);

const { attrsWithoutStyles } = useAttrsWithoutStyles();
</script>

<template>
  <div class="bcc-input-container" :class="$attrs['class']" :style="$attrs['style'] as StyleValue">
    <label class="bcc-input-label" :for="id" v-if="label || showOptionalLabel">
      <span>{{ label }}</span>
      <span v-if="showOptionalLabel" class="bcc-input-optional-label">
        {{ optionalLabel }}
      </span>
    </label>
    <select
      class="bcc-select"
      :class="{
        'bcc-select-error': state === 'error',
        'bcc-select-success': state === 'success',
        'bcc-select-sm': size === 'sm',
        'bcc-select-lg': size === 'lg',
      }"
      :id="id"
      :value="modelValue"
      :disabled="disabled"
      :required="required"
      @input="emit('update:modelValue', ($event.target as HTMLSelectElement).value)"
      v-bind="attrsWithoutStyles"
    >
      <slot />
    </select>

    <span
      v-if="$slots.message"
      :class="{
        'bcc-input-message-default': state === 'default',
        'bcc-input-message-error': state === 'error',
        'bcc-input-message-success': state === 'success',
      }"
    >
      <slot name="message" />
    </span>
  </div>
</template>
