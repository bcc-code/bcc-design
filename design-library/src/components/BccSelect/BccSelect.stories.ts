import BccSelect from "./BccSelect.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

/**
 * A `select` element supporting most of the props of the Input component
 */
export default {
  title: "Forms/BccSelect",
  component: BccSelect,
  argTypes: {
    size: {
      options: ["sm", "base", "lg"],
      control: { type: "radio" },
    },
    state: {
      description: "Style of the select",
      options: ["default", "error", "success"],
      control: { type: "radio" },
    },
    showOptionalLabel: {
      description: "Will only take effect when `required` is `false`",
    },
  },
} as Meta<typeof BccSelect>;

const Template: StoryFn<typeof BccSelect> = (args) => ({
  components: { BccSelect },
  setup() {
    return { args };
  },
  template: `
    <BccSelect v-bind="args" class="w-1/4">
      <option disabled value="">Select an option...</option>
      <option value="one">Option 1</option>
      <option value="two">Option 2</option>
    </BccSelect>
  `,
});

/**
 * Use `v-model` to control the value of the select. Pass normal HTML `option` elements to the default slot to render the options (be sure to include `value` attributes on each option).
 */
export const Example = Template.bind({});
Example.args = {
  modelValue: "one",
  label: "Example label",
  state: "default",
  size: "base",
  disabled: false,
  required: false,
  showOptionalLabel: false,
  optionalLabel: "Optional",
};
Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccSelect
  label="Example label"
  v-model="example"
>
  <option disabled value="">Select an option...</option>
  <option value="one">Option 1</option>
  <option value="two">Option 2</option>
</BccSelect>
`,
    },
  },
};

/**
 * Set the `state` prop to control how the select is rendered. Set the `disabled` prop to disable the select
 */
export const State: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect class="w-1/4" value="default">
        <option value="default">Default</option>
        <option value="two">Option 2</option>
      </BccSelect>
      <BccSelect class="w-1/4" value="default" disabled>
        <option value="default">Default disabled</option>
        <option value="two">Option 2</option>
      </BccSelect>
      <BccSelect state="error" class="w-1/4" value="default">
        <option value="default">Error</option>
        <option value="two">Option 2</option>
      </BccSelect>
      <BccSelect state="error" class="w-1/4" value="default" disabled>
        <option value="default">Error disabled</option>
        <option value="two">Option 2</option>
      </BccSelect>
      <BccSelect state="success" class="w-1/4" value="default">
        <option value="default">Success</option>
        <option value="two">Option 2</option>
      </BccSelect>
      <BccSelect state="success" class="w-1/4" value="default" disabled>
        <option value="default">Success disabled</option>
        <option value="two">Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * Like a normal select element, don't set the value and create an `option` with `disabled` and an empty value to have a default placeholder text.
 */
export const Placeholder: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect class="w-1/4" value="">
        <option disabled value="">Select an option...</option>
        <option value="one">Success</option>
        <option value="two">Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * Set the `size` prop to control the size of the select
 */
export const Size: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect label="label for sm select" size="sm" class="w-1/4" value="default">
        <option value="default">sm select</option>
        <option value="two">Option 2</option>
      </BccSelect>
      <BccSelect label="label for base select" size="base" class="w-1/4" value="default">
        <option value="default">base select</option>
        <option value="two">Option 2</option>
      </BccSelect>
      <BccSelect label="label for lg select" size="lg" class="w-1/4" value="default">
        <option value="default">lg select</option>
        <option value="two">Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * The select can have a `label`. A unique id is automatically generated for the select, which the label uses in its `for` attribute
 */
export const WithLabel: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect label="Example Label" size="base" value="default">
        <option value="default">base select</option>
        <option value="two">Option 2</option>
      </BccSelect>
      <BccSelect label="Example Label that is really long but should still display"  class="w-1/4" value="default">
        <option value="default">base select</option>
        <option value="two">Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * Set the `show-optional-label` prop to show a label when the select is not `required`. Control the text for this label with the `optionalLabel` prop, which can be useful for translation.
 */
export const Optional: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect show-optional-label label="Example Label" class="w-1/4" value="default">
        <option value="default">With label and optional</option>
        <option value="two">Option 2</option>
      </BccSelect>

      <BccSelect show-optional-label label="Example Label" optionalLabel="Optional label that is pretty long and should truncate" class="w-1/4" value="default">
        <option value="default">Long optional label</option>
        <option value="two">Option 2</option>
      </BccSelect>

      <BccSelect show-optional-label optionalLabel="Optional label that is pretty long and should truncate" class="w-1/4" value="default">
        <option value="default">Without label</option>
        <option value="two">Option 2</option>
      </BccSelect>

      <BccSelect show-optional-label label="Pretty long label as well" optionalLabel="Optional label that is pretty long and should truncate" class="w-1/4" value="default">
        <option value="default">Long label and optional label</option>
        <option value="two">Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * Pass text to the `message` slot for the component to render a message below the select. The styling takes the `state` into account.
 */
export const WithMessage: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect class="w-1/2" value="default">
        <option value="default">Default</option>
        <option value="two">Option 2</option>

        <template #message>
          This is an example message with <strong>styling</strong>.
        </template>
      </BccSelect>

      <BccSelect class="w-1/2" state="success" value="default">
        <option value="default">Success</option>
        <option value="two">Option 2</option>

        <template #message>
          This is an example message with <strong>styling</strong>.
        </template>
      </BccSelect>

      <BccSelect class="w-1/2" state="error" value="default">
        <option value="default">Error</option>
        <option value="two">Option 2</option>

        <template #message>
          This is an example message with <strong>styling</strong>.
        </template>
      </BccSelect>
    </div>
  `,
});
