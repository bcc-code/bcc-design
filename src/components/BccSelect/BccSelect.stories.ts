import BccSelect from "./BccSelect.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

/**
 * A `select` element supporting most of the props of the Input component
 */
export default {
  title: "Forms/BccSelect",
  component: BccSelect,
  argTypes: {
    size: {
      options: ["sm", "base", "lg"],
      control: { type: "radio" },
    },
  },
} as Meta<typeof BccSelect>;

const Template: StoryFn<typeof BccSelect> = (args) => ({
  components: { BccSelect },
  setup() {
    return { args };
  },
  template: `
    <BccSelect v-bind="args" class="w-1/4">
      <option>Option 1</option>
      <option>Option 2</option>
    </BccSelect>
  `,
});

/**
 * Use `v-model` to control the value of the select. Pass normal HTML `option` elements to the default slot to render the options.
 */
export const Example = Template.bind({});
Example.args = {
  label: "Example label",
  state: "default",
  size: "base",
  disabled: false,
  required: false,
  showOptionalLabel: false,
  optionalLabel: "Optional",
};
Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccSelect
  label="Example label"
  v-model="example"
>
  <option>Option 1</option>
  <option>Option 2</option>
</BccSelect>
`,
    },
  },
};

/**
 * Set the `state` prop to control how the select is rendered. Set the `disabled` prop to disable the select
 */
export const State: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect class="w-1/4">
        <option>Default</option>
        <option>Option 2</option>
      </BccSelect>
      <BccSelect state="error" class="w-1/4">
        <option>Error</option>
        <option>Option 2</option>
      </BccSelect>
      <BccSelect state="success" class="w-1/4">
        <option>Success</option>
        <option>Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * Set the `size` prop to control the size of the select
 */
export const Size: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect size="sm" class="w-1/4">
        <option>sm select</option>
        <option>Option 2</option>
      </BccSelect>
      <BccSelect size="base" class="w-1/4">
        <option>base select</option>
        <option>Option 2</option>
      </BccSelect>
      <BccSelect size="lg" class="w-1/4">
        <option>lg select</option>
        <option>Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * The input can have a `label`. A unique id is automatically generated for the input, which the label uses in its `for` attribute
 */
export const WithLabel: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect label="Example Label" size="base">
        <option>base select</option>
        <option>Option 2</option>
      </BccSelect>
      <BccSelect label="Example Label that is really long but should still display"  class="w-1/4">
        <option>base select</option>
        <option>Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * Set the `show-optional-label` prop to show a label when the input is not `required`. Control the text for this label with the `optionalLabel` prop, which can be useful for translation.
 */
export const Optional: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect show-optional-label label="Example Label" class="w-1/4">
        <option>With label and optional</option>
        <option>Option 2</option>
      </BccSelect>

      <BccSelect show-optional-label label="Example Label" optionalLabel="Optional label that is pretty long and should truncate" class="w-1/4">
        <option>Long optional label</option>
        <option>Option 2</option>
      </BccSelect>

      <BccSelect show-optional-label optionalLabel="Optional label that is pretty long and should truncate" class="w-1/4">
        <option>Without label</option>
        <option>Option 2</option>
      </BccSelect>

      <BccSelect show-optional-label label="Pretty long label as well" optionalLabel="Optional label that is pretty long and should truncate" class="w-1/4">
        <option>Long label and optional label</option>
        <option>Option 2</option>
      </BccSelect>
    </div>
  `,
});

/**
 * Pass text to the `message` slot for the component to render a message below the select. The styling takes the `state` into account.
 */
export const WithMessage: StoryFn<typeof BccSelect> = () => ({
  components: { BccSelect },
  template: `
    <div class="flex flex-col space-y-4">
      <BccSelect class="w-1/2">
        <option>Default</option>
        <option>Option 2</option>

        <template #message>
          This is an example message with <strong>styling</strong>.
        </template>
      </BccSelect>

      <BccSelect class="w-1/2" state="success">
        <option>Success</option>
        <option>Option 2</option>

        <template #message>
          This is an example message with <strong>styling</strong>.
        </template>
      </BccSelect>

      <BccSelect class="w-1/2" state="error">
        <option>Error</option>
        <option>Option 2</option>

        <template #message>
          This is an example message with <strong>styling</strong>.
        </template>
      </BccSelect>
    </div>
  `,
});
