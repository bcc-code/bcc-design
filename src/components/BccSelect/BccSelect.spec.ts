import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import BccSelect from "./BccSelect.vue";

describe("BccSelect", () => {
  it("emits update event on input", async () => {
    const wrapper = mount(BccSelect, {
      slots: { default: "<option value='brunstad'>Brunstad</option>" },
    });
    await wrapper.find(".bcc-select").setValue("brunstad");

    expect(wrapper.emitted("update:modelValue")?.length).toBe(1);
    expect(wrapper.emitted("update:modelValue")![0]).toEqual(["brunstad"]);
  });

  it("renders a text from the default and message slots", () => {
    const wrapper = mount(BccSelect, {
      slots: {
        default: "<option value='brunstad'>Brunstad</option>",
        message: "Test message",
      },
    });

    expect(wrapper.text()).toContain("Test message");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders a label", () => {
    const wrapper = mount(BccSelect, {
      props: { label: "Test label" },
    });

    expect(wrapper.text()).toBe("Test label");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders without a label or optional text", () => {
    const wrapper = mount(BccSelect);

    expect(wrapper.text()).toBe("");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("can render an extra label if the input is optional", () => {
    const wrapper = mount(BccSelect, {
      props: { showOptionalLabel: true, optionalLabel: "Custom optional label" },
    });

    expect(wrapper.text()).toBe("Custom optional label");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("does not render a text if the input is both required and optional", () => {
    const wrapper = mount(BccSelect, {
      props: { showOptionalLabel: true, optionalLabel: "Custom optional label", required: true },
    });

    expect(wrapper.text()).not.toBe("Custom optional label");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("passes through non-prop attributes", () => {
    const wrapper = mount(BccSelect, { attrs: { name: "test" } });
    expect(wrapper.html()).toContain('name="test"');
  });
});
