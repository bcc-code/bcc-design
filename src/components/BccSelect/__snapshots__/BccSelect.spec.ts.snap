// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccSelect > can render an extra label if the input is optional 1`] = `
"<div class=\\"bcc-input-container\\"><label class=\\"bcc-form-label\\" for=\\"bcc-input-5\\"><span></span><span class=\\"bcc-form-label-optional\\">Custom optional label</span></label><select class=\\"bcc-select\\" id=\\"bcc-input-5\\"></select>
  <!--v-if-->
</div>"
`;

exports[`BccSelect > does not render a text if the input is both required and optional 1`] = `
"<div class=\\"bcc-input-container\\">
  <!--v-if--><select class=\\"bcc-select\\" id=\\"bcc-input-6\\" required=\\"\\"></select>
  <!--v-if-->
</div>"
`;

exports[`BccSelect > renders a label 1`] = `
"<div class=\\"bcc-input-container\\"><label class=\\"bcc-form-label\\" for=\\"bcc-input-3\\"><span>Test label</span>
    <!--v-if-->
  </label><select class=\\"bcc-select\\" id=\\"bcc-input-3\\"></select>
  <!--v-if-->
</div>"
`;

exports[`BccSelect > renders a text from the default and message slots 1`] = `
"<div class=\\"bcc-input-container\\">
  <!--v-if--><select class=\\"bcc-select\\" id=\\"bcc-input-2\\">
    <option value=\\"brunstad\\">Brunstad</option>
  </select><span class=\\"bcc-form-message bcc-form-message-default\\">Test message</span>
</div>"
`;

exports[`BccSelect > renders without a label or optional text 1`] = `
"<div class=\\"bcc-input-container\\">
  <!--v-if--><select class=\\"bcc-select\\" id=\\"bcc-input-4\\"></select>
  <!--v-if-->
</div>"
`;
